clash_lib/src/proxy/tproxy/iptables.rs

```diff
@@ -0,0 +1,219 @@
+            TProxyStrategy::Iptables
+        } else {
+            TProxyStrategy::None
+        }
+    }
+}
+#[derive(Debug)]
+pub struct TProxyGuard {
+    strategy: TProxyStrategy,
+    skip_mark: u32,
+    tproxy_port: u16,
+    output_chain_name: Option<String>,
+    prerouting_chain_name: Option<String>,
+}
+
+impl TProxyGuard {
+    pub fn new(
+        strategy: TProxyStrategy,
+        skip_mark: u32,
+        tproxy_port: u16,
+        output_chain_name: Option<String>,
+        prerouting_chain_name: Option<String>,
+    ) -> Self {
+        strategy.setup(
+            skip_mark,
+            tproxy_port,
+            output_chain_name.as_deref(),
+            prerouting_chain_name.as_deref(),
+        );
+        Self {
+            strategy,
+            skip_mark,
+            tproxy_port,
+            output_chain_name,
+            prerouting_chain_name,
+        }
+    }
+}
+
+impl Drop for TProxyGuard {
+    fn drop(&mut self) {
+        match self.strategy {
+            TProxyStrategy::Iptables => {
+                clean_iptables_tproxy(
+                    self.output_chain_name.as_deref(),
+                    self.prerouting_chain_name.as_deref(),
+                );
+            }
+            TProxyStrategy::None => {}
+        }
+    }
+}
+
+fn command_exists(command: &str) -> bool {
+    Command::new(command).arg("-v").output().is_ok()
+}
+
+const DEFAULT_OUTPUT_DIVERT_CHAIN: &str = "CLASH_TPROXY_OUTPUT_DIVERT";
+const DEFAULT_OUTPUT_CHAIN: &str = "CLASH_TPROXY_OUTPUT";
+const DEFAULT_PREROUTING_CHAIN: &str = "CLASH_TPROXY_PREROUTING";
+const POLICY_ROUTING_TABLE_NUM: u32 = 400;
+const DEFAULT_TPROXY_MARK: u32 = 0x1;
+const DEFAULT_TPROXY_MARK_MUSK: &str = "0x1/0x1";
+
+// TODO: add a cli to mamange these reserved addresses more easily?
+const RESERVED_ADDRS: &[&str] = &[
+    "0/8",
+    "10/8",
+    "100.64/10",
+    "127/8",
+    "169.254/16",
+    "172.16/12",
+    "192/24",
+    "192.0.2.0/24",
+    "192.88.99/24",
+    "192.168/16",
+    "198.18/15",
+    "198.51.100/24",
+    "203.0.113/24",
+    "224/4",
+    "240/4",
+    "255.255.255.255/32",
+];
+
+// TODO: handle corner cases in iptables' rules
+pub fn setup_iptables_tproxy(
+    skip_mark: u32,
+    tproxy_port: u16,
+    output_chain_name: Option<&str>,
+    prerouting_chain_name: Option<&str>,
+) {
+    let divert_chain_name = DEFAULT_OUTPUT_DIVERT_CHAIN;
+    let output_chain_name = output_chain_name.unwrap_or(DEFAULT_OUTPUT_CHAIN);
+    let prerouting_chain_name = prerouting_chain_name.unwrap_or(DEFAULT_PREROUTING_CHAIN);
+
+    debug!(
+        "setup tproxy via iptables, policy_routing_table_num: {}, tproxy_mark: {}, output_chain_name: {}, prerouting_chain_name: {}",
+        POLICY_ROUTING_TABLE_NUM, DEFAULT_TPROXY_MARK, output_chain_name, prerouting_chain_name
+    );
+
+    run_cmd!(ip rule add fwmark $DEFAULT_TPROXY_MARK lookup $POLICY_ROUTING_TABLE_NUM);
+    run_cmd!(ip route add local "0.0.0.0/0" dev lo table $POLICY_ROUTING_TABLE_NUM);
```

Comment:
this would be lowest priority, is it necessary/useful to do what openvpn does? https://serverfault.com/a/312977 (i think mainly for having higher priority than the system default route)

and do we need to delete this route in clean up?